src/rulesets-new/lifecycle-rules.ts

```diff
@@ -0,0 +1,112 @@
+              ", ",
+            )}`,
+          });
+        }
+      },
+    );
+  },
+});
+
+const stabilityTransitions = new SpecificationRule({
+  name: "resource stability transitions",
+  docsLink: links.versioning.promotingStability,
+  rule: (specificationAssertions) => {
+    specificationAssertions.changed(
+      "not change unless it was wip",
+      (before, after) => {
+        const beforeStability: string | undefined = before.value[stabilityKey];
+        const afterStability: string | undefined = after.value[stabilityKey];
+        // When there is no `before`, it means it's a new file. Any stability is allowed.
+        // When there is no `after`, it means it's been deleted and sunset rules should apply in another rule.
+        if (!beforeStability || !afterStability) return;
+        // A resource can go from wip to anything, so no need to check.
+        if (beforeStability === "wip") return;
+
+        if (beforeStability !== afterStability) {
+          throw new RuleError({
+            message: `stability transition from '${before}' to '${after}' not allowed`,
+          });
+        }
+      },
+    );
+  },
+});
+
+const followSunsetRules = new SpecificationRule({
+  name: "sunset rules",
+  matches: (specification, rulesContext) => {
+    return (
+      !(
+        specification.value[stabilityKey] === "wip" ||
+        specification.value[stabilityKey] === "experimental"
+      ) && rulesContext.specification.change === "removed"
+    );
+  },
+  rule: (specificationAssertions, rulesContext) => {
+    specificationAssertions.requirement("follow sunset rules", () => {
+      const {
+        resourceVersions,
+        changeResource: resourceName,
+        changeVersion: { date, stability },
+        changeDate,
+      } = rulesContext.custom;
+
+      const deprecatedBy =
+        resourceVersions?.[resourceName]?.[date]?.[stability]?.deprecatedBy;
+
+      if (!deprecatedBy) {
+        throw new RuleError({
+          message: `expected ${resourceName} to be deprecated before removing`,
+        });
+      }
+
+      const contextDate = new Date(changeDate);
+      const resourceDate = new Date(date);
+
+      const diffDays =
+        (contextDate.getTime() - resourceDate.getTime()) /
+        (1000 * 60 * 60 * 24);
+
+      // Number of days required before a resource can be removed
+      const sunsetSchedule = {
+        beta: 90,
+        ga: 180,
+      };
+      const requiredDays = sunsetSchedule[stability];
+
+      if (!requiredDays) {
+        throw new RuleError({
+          message: `unexpected stability ${stability} in ${resourceName}`,
+        });
+      }
+      if (diffDays < requiredDays) {
+        throw new RuleError({
+          message: `expected ${stability} resource ${resourceName} to be deprecated ${requiredDays} days`,
+        });
+      }
+    });
+  },
+});
+
+export const lifecycleRuleset = new Ruleset({
```

Comment:
Could I register this with the same kind of call as the old rulesets, `snykRulesService.useDslWithNamedRules`?

Reply:
I'm also wondering, how might we select different rulesets from the CLI by name when running compare or bulk-compare? This next cycle, I'm going to need to add a ruleset for linting compiled OpenAPI for gating deployments. Most of the rules can be reused, but a few will be different.

Reply:
Few questions:
- How are you invoking different rulesets? There are a few ways I have in mind we could do this (we could set env variables, and parse them to apply different rulesets in `src/index.ts`
- When is this next cycle / when are you planning on starting? Are you going to use these new rule runner? I've got a few more PRs to throw up but I should have this up tuesday afternoon, then we can do some testing + regression tests, etc and make sure this works e2e (just want to know if your wanting to use this earlier)

Reply:
The next cycle starts officially Wednesday. An environment variable would work, if not a command-line switch.

Reply:
Another option might be an alternative main entry point?

Reply:
> if not a command-line switch

This might be a little harder due to how we initialize and parse the CLI.

Env variable [if statements here](https://github.com/snyk/sweater-comb/blob/main/src/index.ts#L25-L31) would probably be the best option

Something like `process.env.LINT_ENV === 'compiled' ? [...CompiledChecks] : [...NormalChecks]`